// @apps-in-toss/native-modules/src/MiniAppModule/native-modules/getConsentedUserData.ts
import { isMinVersionSupported } from './isMinVersionSupported';
import { safePostMessage } from '../../natives';

/**
 * @public
 * @category 사용자 데이터
 * @name GetConsentedUserDataErrorCode
 * @description `getConsentedUserData` 호출이 실패했을 때 반환될 수 있는 에러 코드예요.
 */
export type GetConsentedUserDataErrorCode =
  | 'USER_DECLINED'
  | 'UNAVAILABLE'
  | 'TERMS_NOT_SET'
  | 'INVALID_REQUEST'
  | 'CANCELED'
  | 'CONSENTED_USER_DATA_AGREEMENT_FAILED'
  | 'CONSENTED_USER_DATA_INVALID_DATA';

/**
 * @public
 * @category 사용자 데이터
 * @name GetConsentedUserDataOptions
 * @description 사용자 동의 기반 데이터를 요청할 때 사용하는 옵션이에요.
 * @property {string} consentedUserDataKey - 서버가 사용자 데이터 제공 동의문과 데이터 묶음을 찾을 때 쓰는 key예요. 예: `cud_delivery`.
 * @property {boolean} [shouldRequestAgreementWhenUserDeclined=false] - `USER_DECLINED` 상태에서도 다시 약관 웹뷰를 띄울지 여부예요. 기본값은 `false`예요.
 */
export interface GetConsentedUserDataOptions {
  consentedUserDataKey: string;
  shouldRequestAgreementWhenUserDeclined?: boolean;
}

type ConsentedUserDataKey =
  | 'USER_NAME'
  | 'USER_GENDER'
  | 'USER_NATIONALITY'
  | 'USER_BIRTHDAY'
  | 'USER_PHONE'
  | 'USER_ADDRESS'
  | 'USER_EMAIL'
  | 'USER_CONSUMPTION_HISTORY';

/**
 * @public
 * @category 사용자 데이터
 * @name ConsentedUserData
 * @description 사용자가 동의한 후 서버에서 제공하는 데이터예요. 요청한 항목에 해당하는 키만 포함될 수 있어요.
 */
export type ConsentedUserData = Partial<Record<ConsentedUserDataKey, string>>;

/**
 * @public
 * @category 사용자 데이터
 * @name getConsentedUserData
 * @description
 * 사용자 동의 기반 데이터를 요청해요.
 * 동의가 필요한 경우 약관 웹뷰를 띄우고, 동의가 완료되면 서버에서 조회한 데이터를 반환해요.
 *
 * @param {GetConsentedUserDataOptions} options - 동의 데이터 요청 옵션이에요.
 * @returns {Promise<ConsentedUserData | undefined>} 서버에서 제공하는 사용자 데이터를 반환해요. 토스앱 버전이 `5.264.0` 미만이면 `undefined`를 반환해요.
 *
 * @throws
 * - {code: `USER_DECLINED`}: 사용자가 명시적으로 거부했고 재요청 옵션이 없거나, 약관 웹뷰에서 동의를 거부한 경우예요.
 * - {code: `UNAVAILABLE`}: 서버가 사용자 동의 여부를 판단할 수 없는 경우예요.
 * - {code: `TERMS_NOT_SET`}: 미니앱에 사용자 데이터 제공 동의문이 설정되지 않은 경우예요.
 * - {code: `INVALID_REQUEST`}: `consentedUserDataKey`가 비어 있거나, `termsUrl`이 없거나, HTTPS 회사 도메인이 아닌 경우예요.
 * - {code: `CANCELED`}: 약관 웹뷰가 결과를 보내기 전에 닫힌 경우예요.
 * - {code: `CONSENTED_USER_DATA_AGREEMENT_FAILED`}: 약관 웹뷰 처리 중 오류가 발생한 경우예요.
 * - {code: `CONSENTED_USER_DATA_INVALID_DATA`}: `PROVIDED`인데 `data`가 없거나, 동의 성공 후 재조회 결과가 `PROVIDED`가 아닌 경우예요.
 *
 * @example
 * ```tsx
 * import { getConsentedUserData } from '@apps-in-toss/framework';
 *
 * async function fetchDeliveryUserData() {
 *   try {
 *     const data = await getConsentedUserData({
 *       consentedUserDataKey: 'cud_delivery',
 *     });
 *
 *     console.log('사용자 이름:', data?.USER_NAME);
 *     console.log('전화번호:', data?.USER_PHONE);
 *   } catch (error) {
 *     console.error('동의 데이터 조회 실패:', error);
 *   }
 * }
 * ```
 */
export async function getConsentedUserData(options: GetConsentedUserDataOptions) {
  const isSupported = isMinVersionSupported({ ios: '5.264.0', android: '5.264.0' });

  if (!isSupported) {
    return undefined;
  }

  return safePostMessage('getConsentedUserData', options);
}
